sqsworkers/crew.py

```diff
@@ -186,6 +195,76 @@ def poll_queue(self):
+                num_not_deleted = len(delete_statuses['Failed']) if 'Failed' in delete_statuses else 0
+                if num_not_deleted > 0:
+                    self.logger.error('Delete unsuccessful for {} out of {} entries'
+                                      .format(num_not_deleted, len(entries)))
+                    for status in delete_statuses['Failed']:
+                        self.crew.statsd.increment('sqs.delete.failure', 1, tags=[])
+                        self.logger.error('delete fail for {}'.format(status))
+            except Exception as e:
+                import traceback
+                traceback.print_exc()
+                # select which exception handler to call based on the argument passed
+                getattr(self, self.exception_handler_function)(e, message)
+                # continue with the next message and do not delete
+                pass
+
+        def clear_processed(processed_list, messages_list):
+            entries = []
+            for processed, message in zip(processed_list, messages_list):
+                if processed is None:
+                    self.logger.debug('Ignoring message {}'.format(message))
+                    continue  # Done to accommodate message processor that returns no status
+                              # and avoid falsely negative statsd
+                if processed:
+                    entries.append({
+                        'Id': message.message_id,
+                        'ReceiptHandle': message.receipt_handle
+                    })
+                    if len(entries) > 9:
+                        delete_from_sqs(entries)
+                        entries.clear()
+                else:
+                    self.crew.statsd.increment('process.record.failure', 1, tags=[])
+            if len(entries) > 0:
+                delete_from_sqs(entries)
+                entries.clear()
+
+        while self.employed:
+            messages = self.queue.receive_messages(
+                AttributeNames=['All'],
+                MessageAttributeNames=['All'],
+                MaxNumberOfMessages=self.max_number_of_messages,
+                WaitTimeSeconds=self.wait_time)
+            if len(messages) > 0:
+                try:
+                    self.logger.info('processing %s messages %s' % (len(messages), messages))
+                    processor = self.crew.MessageProcessor(messages)
+                except Exception as e:
+                    # select which exception handler to call based on the argument passed
+                    for message in messages:
+                        getattr(self, self.exception_handler_function)(e, message)
+                    # continue with the next message and do not delete
+                    pass
+                self.crew.statsd.increment('process.record.start', len(messages), tags=[])
+                processed = processor.start()
```

Comment:
What if the `start` function throws an exception here? We would no longer be able to use the custom exception_handler. :(

Reply:
I see. I will move this inside the try block.